### PR TITLE
FOR_EACH_CLAUSE variable is not declared final

### PR DESCRIPTION
**Pull request by Bhavik3**
...69939c84cd69cd/src/test/resources/com/puppycrawl/tools/checkstyle/coding/InputFinalLocalVariable.java#L134 should be declared final.

    FinalLocalVariableCheck.java is not checking for this variable (because it is not checking for FOR_EACH_CLAUSE variable). following is the line of code:

    https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java#L112

    so, I remove this line. as there is not any conflict with removing this line.

    now there are many places where FOR_EACH_CLAUSE variable is not declared final in source code. so I changed it.

### CODE REVIEW

**Comment by rdiachenko** [comment]
- travis build failed
- code coverage decreased

Please fix

**Comment by Bhavik3** [comment]
it was fine on my local system...

**Comment by rdiachenko** [comment]
@Bhavik3 please remove changes from ```.classpath```

**Comment by Bhavik3** [comment]
Actually, if i remove changes in classpath, it is showing build error for oraclejdk..

**Comment by rdiachenko** [comment]
There are no changes in .classpath that is in master and the build passes fine there. That means something wrong on your side. Could you please investigate?

**Comment by Bhavik3** [comment]
I made the same changes as before when everything was fine on previous PR.
even when i don't change in classpath, working fine on my local system. but not working on Travis build (for oraclejdk).

but, i will look at it after reverting changes in classpath.

**Comment by romani** [comment]
1) please rebase your code to let me merge easily.
2) please update comment of commit to DO NOT HAVE LINK in it.
3) please read http://checkstyle.sourceforge.net/contributing.html#Submitting_your_contribution and please read the whole page - http://checkstyle.sourceforge.net/contributing.html.
4) I do not see changes to Inputs of UTs - any changes in logic should be reflected in Inputs and UTs.
5) Please do functional changes as one commit and changes to whole code base as second commit, I need clearly see changes to Check+Input+UTs.

FYI: I need reply "done" or your comment for each of my point. Nothing should be forgotten or skipped.

**Comment by Bhavik3** [comment]
1>> I  have removed all links in commit.

2>> UTs:
       Actually, there are already, Unit test available for this issue. so, I just update FinalLocalVariableCheck accordingly. no need of additional UTs to add.

3>> coverage is 100% for FinalLocalVaribaleCheck now.

And,

due to stalled build,  The build has been terminated.
can you re-run travis build or I would close and again pull request so that Travis build re-run automatically.

@romani

**Comment by romani** [comment]
commit message should be single line

I am waiting your reply on each my point , I see reply only 3 items from 5

**Comment by Bhavik3** [comment]
1) please rebase your code to let me merge easily.
>>done

2) please update comment of commit to DO NOT HAVE LINK in it.
>>done

3) please read http://checkstyle.sourceforge.net/contributing.html#Submitting_your_contribution and please read the whole page - http://checkstyle.sourceforge.net/contributing.html.
>> read very carefully.

4) I do not see changes to Inputs of UTs - any changes in logic should be reflected in Inputs and UTs.
>> Already answered in previous comment.

5) Please do functional changes as one commit and changes to whole code base as second commit, I need clearly see changes to Check+Input+UTs.
>>Ruslan told me to squash all my commit into one. so I squashed all my changes into one.
>>If you want: FinalLocalVariableCheck in one commmit
                       input (UTs) into one
                       changes in other related files change in third

Thanks for your support.

**Comment by romani** [comment]
1) please split commit into two, one for Check+Uts+Input, second for whole Checkstyle code to follow new Check requirements. It is very hard to do code review on all changes combined. 
Ruslan might mean to have one commit for all functional changes.


2) 
> FOR_EACH_CLAUSE variable is not reported fixed.reference #20

please make it "fix for FinalLocalVariableCheck - FOR_EACH_CLAUSE variable is not reported. issue #20"

**Comment by Bhavik3** [comment]
done 
@romani

**Comment by romani** [comment]
Message for commit should describe update in it, so second commit should be renamed to describe its changes but still have reference to issue.

Please keep reply me "done " to each my point.

**Comment by Bhavik3** [comment]
done
@romani

**Comment by romani** [comment]
what was done ?

**Comment by Bhavik3** [comment]
I just changed in commit, nothing else in file.
Also working fine on local system.

**Comment by romani** [comment]
I still see (on gihub) two commits named the same way

**Comment by Bhavik3** [comment]
I appended at end. for what this commit is (UTs / Check/ Input).

**Comment by romani** [comment]
I see two equal meassages - https://github.com/Bhavik3/checkstyle/commits/issue20
messaged should be significantly different as in second commit you do not fix Check at all.

**Comment by romani** [comment]
1) InputFinalLocalVariable.java

> public isAbstract(final int x){

What is a reason of this update ?
If not related to issue but just coverage improvement, please move that change to separate PR and explain reason of it separately.

**Comment by Bhavik3** [comment]
ok .... wait

**Comment by Bhavik3** [comment]
Why build is showing error.
it is as same as before except change in commit message

**Comment by romani** [comment]
just open a Travis details and see a reason - https://travis-ci.org/checkstyle/checkstyle/jobs/51409933#L721

here is some Travis problem to download artifacts , I usually restart the build , restarted.

**Comment by Bhavik3** [comment]
again same
https://travis-ci.org/checkstyle/checkstyle/jobs/51409936#L436

it happens to me many times

**Comment by romani** [comment]
if it fail for internet connection reason please to not pay attention ot it, I will restart it before merge. If atleast one Travis build is passed your are ok.

**Comment by Bhavik3** [comment]
ok thanks.

maybe, here coverage will decrease by 0.01% at end.

once this is done,  I will do another PR for coverage problem.

**Comment by romani** [comment]
1) sequence of commit should logical so first commit is changes Check, second id for Checkstyle code base to satisfy "verify" phase of maven build

2) second commit is "fix for FinalLocalVariableCheck - FOR_EACH_CLAUSE variable is not reported" but where are changes or that Check in it ? Please describe EXACT reason of update in message.

**Comment by romani** [comment]
1) please change message for commit "c77feac" to  "fix for FinalLocalVariableCheck - FOR_EACH_CLAUSE variable is not reported. issue #20" as was requested before. Main idea is that anybody should look at list of commits messages and understand what is inside of that commit without looking in it.


2) So now I see update to Check and I do not like it. Nobody use final in FOR_EACH loops. So after release we will get a request to suppress that violation. The same model is with making as "final" all function arguments - nobody do that as it put extra information in method and it become hard to read it. That problems usually resolved by special validation that catch act of assigning new value argument of function or item of iteration (that is definetely a bad practice).
So your update will not be welcome by java community. What we could do is  to make special option "validateEnhancedForVariable"="true/false" that will report problem on that case or not by use desire (by means of that we will reconsile our user desire to validate that or not). Defaut value should be "false" for this option.

FYI: http://docs.oracle.com/javase/specs/jls/se8/html/jls-14.html#jls-14.14.2

**Comment by Bhavik3** [comment]
test.java :

public class InputFinalLocalVariableCheck
{
    public void method() 
    {
	for(int a : getInts())
        {

        }
    }
}

test.xml : 

<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC
    "-//Puppy Crawl//DTD Check Configuration 1.2//EN"
    "http://www.puppycrawl.com/dtds/configuration_1_2.dtd">

<module name="Checker">
  <module name="TreeWalker">
    <module name="FinalLocalVariable">
      <property name="tokens" value="VARIABLE_DEF,PARAMETER_DEF"/>
      <property name="validateEnhancedForVariable" value="true"/>
    </module>
  </module>
</module>

CLI launch:

Starting audit...
/home/bhavik/Documents/checkstyle-test/test.java:5:17: Variable 'a' should be declared final.
Audit done.

**Comment by Bhavik3** [comment]
done
@romani

**Comment by romani** [comment]
good ,

1) 
> DEFAULT_VALIDATE_ENHANCED_FOR_VARIABLE

please remove that variable, no need to make such variable.

2) please extend javadoc with description for new option and find in xdoc folder (it is used to generate our web site ) description for that check too and update it. 

3) please generate HTML reports against few projects to make sure that no exceptions happen and validation violations are correct (example of how other do that: https://github.com/pirat9600q/checkstyle-tester , https://github.com/pirat9600q/pirat9600q.github.io)

**Comment by Bhavik3** [comment]
Hi,
I am unable to generate HTML report.

This is my checkstyle-tester : 
https://github.com/Bhavik3/checkstyle-tester

Generated report : 
https://github.com/Bhavik3/bhavik3.github.io

In report, 
http://bhavik3.github.io/checkstyle.html

files checked are zero.

@romani

**Comment by rdiachenko** [comment]
Hi @Bhavik3 
- Please do 1) and 2) notes from Roman and mark them as 'done'
- what about 3) note, I've just pulled your branch (https://github.com/Bhavik3/checkstyle-tester/tree/checkstyle_FinalLocalVariableCheck) and generated the report by running ```./launch.sh```. Didn't have problems with that. Can you open the generated site at least locally?

**Comment by Bhavik3** [comment]
1) DONE

2) DONE

3) DONE
         Checkstyle-tester configuration : https://github.com/Bhavik3/checkstyle-tester/tree/checkstyle_FinalLocalVariableCheck
         Checkstyle-tester Report : http://bhavik3.github.io/checkstyle.html

**Comment by rdiachenko** [comment]
1) Provide more tests. Currently you test only one case when ```validateEnhancedForVariable=true``` and there is ```for(Object a : list)``` without ```final```. Update your test inputs as well.

2)
```
@Test
+    public void testEnhancedForVariable() throws IOException, Exception
```

What's the reason of throwing Exception?

3)
```Enhanced For Variable``` change to "enhanced for-loop variable"

4)
```private boolean validateEnhancedForVariable(DetailAST ast)``` - unclear method's name (see example of a good name: isVariableInForInit)

5)
```
return validateEnhancedForVariable
+                || ast.getParent().getType() != TokenTypes.FOR_EACH_CLAUSE;
```

I doubt about this condition. Please update unit tests and check that all code branches are covered.

**Comment by Bhavik3** [comment]
1) By default, validateEnhancedForVariable=false. so I thought no need of checking it separately. but now I added for both separately.
DONE 

2) IOException Removed.

3) DONE

4) DONE

5)

return validateEnhancedForVariable
                || ast.getParent().getType() != TokenTypes.FOR_EACH_CLAUSE;

This condition is working fine on all cases.
Rationale:

validateEnhancedForVariable = true , ast.getParent().getType() = FOR_EACH_CLAUSE
>> condition is true and will check for Enhanced For Variable

validateEnhancedForVariable = true , ast.getParent().getType() != FOR_EACH_CLAUSE
>> condition will true and will check as usual (like no condition is their).

validateEnhancedForVariable = false , ast.getParent().getType() = FOR_EACH_CLAUSE
>> condition is false and will not check for Enhanced For Variable (as expected)

validateEnhancedForVariable = false , ast.getParent().getType() != FOR_EACH_CLAUSE
>> condition is true and will check as usual.

so, It will work fine in all test cases.

@rdiachenko

**Comment by rdiachenko** [comment]
1)
Change ```validation of For_EACH_CLAUSE_varible``` to ```Controls whether to check enhanced for-loop variable.```

2)
Change 

```
 * set the validation of enhanced For_Variable.
+     * @param validateEnhancedForLoopVariable true/false.
```

to

```
Whether to check enhanced for-loop variable or not.
@param validateEnhancedForLoopVariable whether to check for-loop variable
```

3)
Change

```
+     * degtermine weather to check for FOR_EACH_CLAUSE varible.
+     * @param ast The ast to compare.
+     * @return true if validation for enhanced for variable is enable.
+     */
+    private boolean weatherToCheckEnhancedForLoopVariable(DetailAST ast)
```

to

```
+     * Determines whether enhanced for-loop variable can be checked or not.
+     * @param ast The ast to compare.
+     * @return true if enhanced for-loop variable can be checked.
+     */
+    private boolean canCheckEnhancedForLoopVariable(DetailAST ast)
```

4)
Change

```
<td>Controls weather to check Enhanced For Loop Variable.</td>
```

to

```
<td>Controls whether to check enhanced for-loop variable.</td>
```

**Comment by Bhavik3** [comment]
all done
@rdiachenko

**Comment by rdiachenko** [comment]
@romani, I am ok to merge

**Comment by romani** [comment]
1)
> FinalLocalVariableCheck.java

no update for Javadoc is NOT acceptable, one day we will generate xdoc from javadoc

2)
> ``` <td>Controls whether to check enhanced for-loop variable.</td>```

Please make a link to JDK specification for "enhanced for-loop", not all Checkstyle users are that smart to understand that term, even I was surprised on how it is named actually.

3)
> 
```
+ private boolean validateEnhancedForLoopVariable;
+ /**
```
Pelase make empty line between declarations

**Comment by Bhavik3** [comment]
1)
I could not find target file to update javadoc in xdocs folder except config_coding.xml
can you tell me which one is the target file.

2)
DONE

3)
DONE

**Comment by romani** [comment]
This is documentation comment (Javadoc) - https://github.com/Bhavik3/checkstyle/blob/44d05139136a7d42b99007a4d21cbe004e39239a/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java#L31

Here should be all information on how to use Check , see example at good javadoc at https://github.com/Bhavik3/checkstyle/blob/44d05139136a7d42b99007a4d21cbe004e39239a/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/HiddenFieldCheck.java

**Comment by Bhavik3** [comment]
DONE
@romani

**Comment by romani** [comment]
1) Revert changes at line 35

2) please provide human description of option in javadoc amd only after that provide example For it.

**Comment by Bhavik3** [comment]
1) DONE

2) It is for developers, so I did not explain what is Enhanced For-Loop Variable. I just explained about what is option for it and how to use it.
DONE

@romani

**Comment by romani** [comment]
yes it is for developers and in future for xdoc generation so - please do provide example of enhanced FOR and link to specification. Do not expect that checkstyle is used by well educated users.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java [hunk 1]
```diff
@@ -39,6 +39,19 @@
  *     &lt;property name="token" value="VARIABLE_DEF"/&gt;
  * &lt;/module&gt;
  * </pre>
+ * <p>
+ * By default, FinalLocalVariableCheck does not check for Enhanced For-Loop Variable.
+ * To enable it,there is an option called "validateEnhancedForLoopVariable".It takes boolean value.
```
Comment (mkordas) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java]: add space between `.` and `It`

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java [hunk 2]
```diff
@@ -155,6 +180,17 @@ public void visitToken(DetailAST ast)
     }
 
     /**
+     * Determines whether enhanced for-loop variable can be checked or not.
+     * @param ast The ast to compare.
+     * @return true if enhanced for-loop variable can be checked.
+     */
+    private boolean canCheckEnhancedForLoopVariable(DetailAST ast)
```
Comment (mkordas) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java]: I think we always can check this. The question is if we should :) Rename method and Javadoc to `shouldCheckEnhancedForLoopVariable`.

File: src/xdocs/config_coding.xml [hunk 3]
```diff
@@ -344,6 +344,19 @@ number.equals(i + j);
               href="apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#VARIABLE_DEF">VARIABLE_DEF</a>
             </td>
           </tr>
+          <tr>
+            <td>validateEnhancedForLoopVariable</td>
+            <td>Controls whether to check<a href = "http://docs.oracle.com/javase/specs/jls/se7/html/jls-14.html#jls-14.14.2"> enhanced for-loop </a>variable.</td>
```
Comment (mkordas) [inline src/xdocs/config_coding.xml]: Please remove obsolete spaces from link. Change `> enhanced for-loop <` to `>enhanced for-loop<` and put spaces outside.

Comment (mkordas) [inline src/xdocs/config_coding.xml]: Maybe it would be better to give link to Java 8 JLS?

**Comment by Bhavik3** [comment]
1)
@mkordas 
Thanks for correcting me.
DONE

2)
@romani 
JAVADOC updated.
DONE

**Comment by romani** [comment]
> Following shows how enhanced For-Loop works for Arrays.

amke javadoc to looks like:

"By default this Check skip final validation on <enhanced For-Loop><LINK>
Option 'validateEnhancedForLoopVariable' could be used to make Check to validate even variable from Enhanced For Loop.
<CONFIG>

 for (int number : myNumbers) { // violation
   System.out.println(number);
 }
"

**Comment by Bhavik3** [comment]
@romani 
DONE

**Comment by romani** [comment]
Good, now apply all changes in javadoc to xdoc, as user read results of xdoc and still do not have tool to generate xdoc from javadoc.

**Comment by Bhavik3** [comment]
u mean in config_coding.xml ?
@romani

**Comment by romani** [comment]
You have only changes to option, but all changes in javadoc should appear in xml

**Comment by Bhavik3** [comment]
DONE
@romani

**Comment by romani** [comment]
you did it !
thanks a lot !!!

**Pull request merged**